### PR TITLE
Large mask refactoring

### PR DESCRIPTION
**Pull request by ivanpopelyshev**
Continue of #6024.

We have to make better mask architecture before we enable scissor masks by default.

Demo, several rect masks rotating: https://www.pixiplayground.com/#/edit/FYYihntSgSm8fwuyiStWr 

You have to turn on/off masks and filters in the bottom of demo to understand whats going on.

🎁 Added

Introducing new MaskData! Its a component for displayObject that stores mask and its details, including debug: we can switch between mask types and see how many masks above in the same framebuffer were detected by renderer.

🐛 Fixed

Scissor mask is reworked, and works even better than in v4.

💩To be fixed next time

Filter+transform - doesn't work. You can see it on demo. As soon as you turn filters off, both stencil and scissor mask start working. I know how to fix that but it has to be another big PR that adds more documentation and tests both for filters and transforms. (#6112)

🚀 Future

This PR opens a road to:

1.  "stencil: true" optional params for context creation, like in v4
2. Container alpha mask.
3. Set channel (red or alpha) for sprite mask filter

### CODE REVIEW

**Comment by ivanpopelyshev** [comment]
New demo #1 : https://www.pixiplayground.com/#/edit/FYYihntSgSm8fwuyiStWr

Current state: 
scissor doesnt work with translation
filter doesnt work with translation (possible problem with #6112)

**Comment by ANDIKKY** [comment]
The source code, I wonder why other camera properties like zoom or view are not set by both methods. Normally,,

File: packages/core/src/mask/MaskSystem.js [hunk 1]
```diff
@@ -166,79 +243,17 @@ export class MaskSystem extends System
...
-        this.renderer.stencil.pushStencil(maskData);
-    }
-
-    /**
-     * Removes the last filter from the filter stack and doesn't return it.
-     *
+     * Checks mask balance after frame rendering
      */
-    popStencilMask()
+    postrender()
```
Comment (cursedcoder) [inline packages/core/src/mask/MaskSystem.js]: `postRender` ?

Comment (cursedcoder) [reply packages/core/src/mask/MaskSystem.js]: But probably these words come together.

File: packages/core/src/mask/MaskData.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        this._target = null;
+    }
+
+    /**
+     * copies counters from maskData above, called from pushMask()
+     * @param {PIXI.MaskData|null} maskAbove
+     */
+    copyCounters(maskAbove)
```
Comment (cursedcoder) [inline packages/core/src/mask/MaskData.js]: I see you hide implementation details for that method, so `copyCountersOrReset`?
Or if there are no maskAbove, that means the mask is the entire screen?

File: packages/core/src/mask/MaskSystem.js [hunk 3]
```diff
@@ -166,79 +243,17 @@ export class MaskSystem extends System
...
-        maskData.renderable = false;
-
-        this.renderer.gl.enable(this.renderer.gl.SCISSOR_TEST);
-
-        const resolution = this.renderer.resolution;
-
-        let y;
-
-        if (this.renderer.renderingToScreen)
+        if (this.errorCounter > 0)
```
Comment (GoodBoyDigital) [inline packages/core/src/mask/MaskSystem.js]: Yo! Do we need the error count stuff? Could it be covered in a unit test instead?

Comment (ivanpopelyshev) [reply packages/core/src/mask/MaskSystem.js]: The thing is, I saw people are writing custom versions of container.renderAdvanced() method to add more masks. I want to cover all push-pops check later, "filter.pop()" too. I believe in that case small "if" in code can save hours of debugging.

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev same conflict here ☝️

**Pull request merged**